### ISSUE TITLE
Mean-Variance-Standard Deviation Calculator (Certification Project)

### ISSUE THREAD

**Issue by beaucarnes** [freeCodeCamp/CurriculumExpansion#236]
Create project from the Python data analysis certification.

**Comment by scissorsneedfoodtoo** [comment]
Project prototype: https://repl.it/@BeauCarnes/fcc-mean-var-std

**Comment by ArieRS** [comment]
https://repl.it/repls/BitesizedOutstandingWheel

Make the readme with point bullet so it is easier to understand
It is easy for people that have knowledge in unit test so he knows what program expected...

**Comment by rlabuonora** [comment]
Hi! Here are a few thoughts on the Mean-Variance-Standard deviation project (it's my first time giving feedback so excuse me if I'm on the wrong track):

1) Instructions:

> Create a function named calculate() in mean_var_std.py that uses Numpy to output the mean, variance, and standard deviation of a 3 x 3 matrix.

I'm not sure a matrix has a mean, a standard deviation and a variance. I would rephrase it to something like: 

> ... output the mean, standard deviation and variance of the rows, the columns, and a vector with all the element of a 3 x 3 matrix.

2) If the list is not length 9, I would raise an error instead of returning a string with an error message.

3) I mostly got this generic error message when tests were failing:

```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Some more specific tests I can think of would test: 

- The keys of the output dictionary `"mean",` `"variance"`, `"standard deviation"`.
- The values of the output dictionary (check that the user returns lists and not `np.array`, etc.
- You should probably use `assertAlmostEqual` instead of `assertEqual` for double comparison, that will make the tests less brittle. [This might be helpfull](https://stackoverflow.com/questions/8929005/unittest-sometimes-fails-because-floating-point-imprecision)

4) Here is my solution, in case you are interested:

```
def calculate(lst):
  import numpy as np

  if len(lst) != 9:
    return "List must contain nine numbers."
  x = np.array(lst).reshape(3, 3)
  result = {
    k: [func(x, axis=ax).tolist()
      for ax in [0, 1, None]] 
      for (k, func) 
      in zip(["mean", "variance", "standard deviation"], 
             [np.mean, np.var, np.std])
  } 


  return result
```

The project is quite nice, I hope this feedback is useful!

**Comment by scissorsneedfoodtoo** [comment]
Awesome, thank you for your feedback @ArieRS and @rlabuonora. We'll work on making the description easier to read with bullet points, and adjust the wording so the goal is more clear. Also, thank you for the suggestions about how to improve the tests and for posting your solution. Both will be very helpful going forward.

**Comment by arthigos** [comment]
Hello!
This is the first alpha project I tried.
I'll get better at generating feedback as I go through the other assignments.

### Feedback
The README.md specification was clear.
Without having much background in Numpy, I watched the FCC YT Video 'Python Numpy Tutorial for Beginners' and was able to get a hang of the syntax.
The troubles I ran into were related to array dimensions, axis, axis indexing, and input reshaping.
I went through the reference manuals available at scipy.org and numpy.org to find the methods needed (reshape, array, tolist(), flatten()).
One thing that got me confused on the README.md was the axis index for the calculations...
Using axis = 1 and axis = 2, I got a tuple index out of range.
When I assumed axis = 0 and axis = 1, the function worked and passed the test.

### Questions
1. The flattened matrix should have which shape? (1,9) or (9,) or other?
2. Should I log the output of my run attempts to catch any unexpected behavior and enrich my feedback?

https://repl.it/@arthigos/fcc-mean-var-std

**Comment by allocthon** [comment]
Fun little problem.  I think the only thing that wasn't clear to me is which axis to use the different operators on first..  
e.g.
np.mean(arr[0,:]) vs np.mean(arr[:,0])

other than that the problem was quite clear.  

I used **np.array**, **np.mean**, **np.var**, **np.std**, and **.reshape** methods.  

Also, this problem might be a nice opportunity to state that you can use different axes, or use the .T **tranpose** method and be lazy..

e.g.  np.mean(arr.T[0,:]) is equivalent to np.mean(arr[:,0])

**Comment by borntofrappe** [comment]
Completed project: https://repl.it/@borntofrappe/fcc-mean-var-std

---

I had no issues with the README, and as a first-timer with Numpy, this was a nice project to get acclimated with the library.

There's a small typo where the sentence describes the length of the list (emphasis mine)

> If a list **contianing** less than 9 elements is passed into the function,

I've misspelled _contains_ so often, that I pay extra attention to the verb.

**Comment by scissorsneedfoodtoo** [comment]
Thank you for your feedback @allocthon. We will try to update the instructions to make that more clear. And thank you for catching that typo @borntofrappe :+1: :+1:

**Comment by beaucarnes** [comment]
I've implemented a lot of the feedback. Thanks everybody.

I used suggestion of @rlabuonora and changed the requirement to raising an exception instead of printing a message when there not 9 numbers in the array.

**Comment by adamdune** [comment]
I don't have much to add, it's a nice little project. I probably should've tried dedicating more time understanding `numpy` before jumping in. 👍 

I was surprised to find the solution in `mean_var_std.py`, perhaps you guys accidentally left it there. 

Also, the unit test raises a deprecationWarning (`test_module.py` line 13)
```
./home/runner/fruc3xe6am9/test_module.py:13: DeprecationWarning: Please use assertRaisesRegex instead.
  self.assertRaisesRegexp(ValueError, "List must contain nine numbers.", mean_var_std.calculate, [2,6,2,8,4,0,1,])
.
```

**Comment by beaucarnes** [comment]
@adamdune Thanks, fixed both those issues.

**Comment by darrentriplett** [comment]
This was the first project I tried and I thought the instructions were really clear for the most part.

The only thing I was unsure about was the last instruction:

Submitting
Copy your project's URL and submit it to freeCodeCamp.

When you say submit to freeCodeCamp do you mean here or somewhere else?

It could be helpfult to have a hyperlink that takes you right to the place to submit your feedback.

My solution is below.

https://repl.it/@discohokie/HugePalatableCoin

**Comment by jdelarubia** [comment]
Link to my solution [mean var std](https://repl.it/@jdelarubia/06-fcc-mean-var-std).
No problems found with the problem requirements.

**Comment by scissorsneedfoodtoo** [comment]
@darrentriplett, sorry for the confusion. When these projects are released, campers will submit a URL to their working project like they do for [other required projects](https://www.freecodecamp.org/learn/front-end-libraries/front-end-libraries-projects/build-a-random-quote-machine). That's what those instructions refer to. 

One improvement we could make to the README files is to have a link to each project's page where they should submit said URL. We can do that once we build the pages for each of these projects.

**Comment by casraysa** [comment]
Hi there!

This is the first project I tried and the instructions were really clear. I didn't find any problems to solve this assignment.

My proposed solution code below:
https://repl.it/@casraysa/fcc-mean-var-std

**Comment by manshul1807** [comment]
Done with this project. Pretty easy i must say. I would be interested to know what was the intention of this project ( in a positive way ) . Probably to get familiar with a bit of numpy. It would have been better if some more situations are added in the project. 

Also, I believe there should be error if the length of list is greater than 9.

Link : https://repl.it/@ManshulArora/fcc-mean-var-std#mean_var_std.py

**Comment by borntofrappe** [comment]
I've revisited the project and have a few suggestions for [the most recent README](https://repl.it/@freecodecamp/fcc-mean-var-std#README.md).

- in the introduction, I found the instructions to be redundant. Instead of repeating the requirements of the project, perhaps it's better to describe the first two paragraphs as follows

  > Create a function named `calculate()` in `mean_var_std.py` that uses Numpy to compute descriptive statistics for a 3 x 3 matrix.
  >
  > The input of the function should be a list containing 9 digits. The function should convert the list into a 3 x 3 Numpy array, and then return a dictionary containing the mean, variance, standard deviation, max, min, and sum along both axes and for the flattened matrix.

- following the example, the README should describe how the error message is returned if _other_ than 9 elements are included.

  > If a list containing ~less~ other than 9 elements is passed into the function

- again following the example, it is perhaps better to break the text in two paragraphs, since the two concepts are rather independent of one another

  > If a list containing other than 9 elements is passed into the function, it should raise a `ValueError` exception with the message: "List must contain nine numbers."
  >
  > The values in the returned dictionary should be lists and not Numpy arrays.

**Comment by rafaelmelhem** [comment]
> Done with this project. Pretty easy i must say. I would be interested to know what was the intention of this project ( in a positive way ) . Probably to get familiar with a bit of numpy. It would have been better if some more situations are added in the project.
> 
> Also, I believe there should be error if the length of list is greater than 9.
> 
> Link : https://repl.it/@ManshulArora/fcc-mean-var-std#mean_var_std.py

I wrote my code almost like you but was getting trouble with my arrays, will correct mine using yours as reference :) 
Thanks for sharing!
